### ISSUE TITLE
Incorrect type inference after type narrowing using isinstance with protocols

### ISSUE THREAD

**Issue by luwqz1** [microsoft/pyright#10270]
**Describe the bug**
Type inference does not change when narrowing to a union of concrete and protocol types that define identical method signatures.

**Code or Screenshots**
```python
import typing


@typing.runtime_checkable
class Foo(typing.Protocol):
    def foo(self) -> str: ...


@typing.runtime_checkable
class Bar(typing.Protocol):
    def foo(self) -> str: ...


def func(x: typing.Any) -> None:
    if (is_foo_bar := isinstance(x, Foo | Bar)) or isinstance(x, str):
        typing.assert_type(x.foo() if is_foo_bar else x, str)  # (reportAssertTypeFailure)
```

However, if we change the return type of the method in one of the other classes, everything works correctly:

```python
import typing


@typing.runtime_checkable
class Foo(typing.Protocol):
    def foo(self) -> str: ...


@typing.runtime_checkable
class Bar(typing.Protocol):
    def foo(self) -> int: ...


def func(x: typing.Any) -> None:
    if (is_foo_bar := isinstance(x, Foo | Bar)) or isinstance(x, str):
        typing.assert_type(x.foo() if is_foo_bar else x, str | int)  # OK
```

It seems this issue only occurs with protocols; everything works as expected with no protocol classes

[pyright playground](https://pyright-play.net/?pythonVersion=3.13&strict=true&code=JYWwDg9gTgLgBDAnmYA7A5gKGwASSjAOigFdUZQBTAfQGMALS2gawEMAjAG0s1s9YDOAuADEIEABT406QgAUoEGBFoROASgBcmOLrgATSgDM4R8RIGVOR9XAC0APjgCYUTXEKfsmPMhnEyChAaBiY2Lh4%2BQWEAIVYoKT8iBSUVNS0dPUMTM0lLa1tHZ1d3T0JvbNMyWgkAD3dpIgBBVERCpwA5CFRKbT04YBMJYAFqXOp2eLhNAF4BgTQXVlRaSjqAGlFxOAAfODiodVtoecWYZdWN4sO%2B-r1G2WjKWGp8NdrCXIlbQfmx8QmUysljgtU2LkOugAxHAJFBKJBYE0hM8YAAVZCUESsYCcEjw9RAA)

**Comment by erictraut** [comment]
Pyright is working as intended here, so this isn't a bug.

Types are narrowed for individual expressions (such as `x`), and the type is not conditioned on the types or values of other expressions. Within the `if` block, the type of `x` is narrowed to `Foo | Bar | str` based on the two `isinstance` conditionals. The narrowed type isn't somehow conditioned on `is_foo_bar`. There is no "entanglement" between the narrowed type of `x` and the type or value of `is_foo_bar`.

If you want this code to work with static type checking, you'd need to do something like this:

```python
def func(x: typing.Any) -> None:
    if isinstance(x, Foo | Bar):
        reveal_type(x)  # Foo | Bar
    elif isinstance(x, str):
        reveal_type(x)  # str
```

**Comment by erictraut** [comment]
There's more involved in your code sample than I previously realized. The target of the walrus operator (`is_foo_bar`) is being treated as an [aliased conditional expression](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=aliased-conditional-expression). I forgot that I had added support for the walrus operator in this feature. So my previous analysis was incorrect.

Let's simplify the code to eliminate the aliased conditional expression. I'm also going to eliminate the `Any` since that involves special code paths in the type narrowing code.

```python
from typing import Protocol, runtime_checkable

@runtime_checkable
class Foo(Protocol):
    def foo(self) -> int: ...

@runtime_checkable
class Bar(Protocol):
    def foo(self) -> int: ...

def func(x: Foo | Bar | str) -> None:
    if isinstance(x, Foo | Bar):
        reveal_type(x) # Foo | Bar
    else:
        reveal_type(x) # Foo | Bar | str (expected str)
```

This is not the behavior I'd expect. 

I'm going to reopen the issue and investigate further.
